components/tikv_alloc/src/jemalloc.rs

```diff
@@ -21,20 +27,103 @@ lazy_static! {
+unsafe impl<T: Send> Send for PeekableRemoteStat<T> {}
+
+impl<T: Copy> PeekableRemoteStat<T> {
+    /// Try access the underlying data. When the pointer is `nullptr`, returns
+    /// `None`.
+    ///
+    /// # Safety
+    ///
+    /// The pointer should not be dangling. (i.e. the thread to be traced should
+    /// be accessible.)
+    unsafe fn peek(&self) -> Option<T> {
+        self.0
+            .map(|nlp| unsafe { core::intrinsics::atomic_load_seqcst(nlp.as_ptr()) })
+    }
+
+    fn from_raw(ptr: *mut T) -> Self {
+        Self(NonNull::new(ptr))
+    }
+}
+
+impl PeekableRemoteStat<u64> {
+    fn allocated() -> Self {
+        // SAFETY: it is transparent.
+        // NOTE: perhaps we'd better add something like `as_raw()` for `ThreadLocal`...
+        Self::from_raw(
+            tikv_jemalloc_ctl::thread::allocatedp::read()
+                .map(|x| unsafe { std::mem::transmute(x) })
+                .unwrap_or(std::ptr::null_mut()),
+        )
+    }
+
+    fn deallocated() -> Self {
+        // SAFETY: it is transparent.
+        Self::from_raw(
+            tikv_jemalloc_ctl::thread::deallocatedp::read()
+                .map(|x| unsafe { std::mem::transmute(x) })
+                .unwrap_or(std::ptr::null_mut()),
+        )
+    }
+}
+
 struct MemoryStatsAccessor {
-    // TODO: trace arena, allocated, deallocated. Original implement doesn't
-    // work actually.
+    allocated: PeekableRemoteStat<u64>,
+    deallocated: PeekableRemoteStat<u64>,
     thread_name: String,
 }
 
-pub fn add_thread_memory_accessor() {
+impl MemoryStatsAccessor {
+    fn get_allocated(&self) -> u64 {
+        // SAFETY: once the memory stats accessor has been added to the map,
+        // we registered the `StillAlive` structure for tracing the thread's lifetime.
+        unsafe { self.allocated.peek().unwrap_or_default() }
+    }
+
+    fn get_deallocated(&self) -> u64 {
+        // SAFETY: once the memory stats accessor has been added to the map,
+        // we registered the `StillAlive` structure for tracing the thread's lifetime.
+        unsafe { self.deallocated.peek().unwrap_or_default() }
+    }
+}
+
+/// Register the current thread to the collector that collects the jemalloc
+/// allocation / deallocation info.
+///
+/// Generally you should call this via `spawn_wrapper`s instead of invoke this
+/// directly. The former is a safe function.
+///
+/// # Safety
+///
+/// Make sure the `remove_thread_memory_accessor` is called before the thread
+/// exits.
+pub unsafe fn add_thread_memory_accessor() {
     let mut thread_memory_map = THREAD_MEMORY_MAP.lock().unwrap();
-    thread_memory_map.insert(
-        thread::current().id(),
-        MemoryStatsAccessor {
-            thread_name: thread::current().name().unwrap().to_string(),
-        },
-    );
+    thread_memory_map
+        .entry(thread::current().id())
+        .or_insert_with(|| {
+            let allocated = PeekableRemoteStat::allocated();
```

Comment:
As protected by MEMORY_MAP's lock and getting rid of thread local , is `PeekableRemoteStat` of any use now?

Reply:
Nope, it is just a thin wrapper over a raw pointer that making it a little bit safer.